src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+        /// </summary>
+        private readonly string _temporaryDirectory;
+
+        /// <summary>
+        /// Map of currently open generated files; the key is the generated full file path.
+        /// </summary>
+        private readonly Dictionary<string, OpenSourceGeneratedFile> _openFiles = new Dictionary<string, OpenSourceGeneratedFile>();
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SourceGeneratedFileManager(
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IThreadingContext threadingContext,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            ITextDocumentFactoryService textDocumentFactoryService,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _serviceProvider = serviceProvider;
+            _threadingContext = threadingContext;
+            _textDocumentFactoryService = textDocumentFactoryService;
+            _temporaryDirectory = Path.Combine(Path.GetTempPath(), "VisualStudioSourceGeneratedDocuments");
+            _visualStudioWorkspace = visualStudioWorkspace;
+
+            Directory.CreateDirectory(_temporaryDirectory);
+
+            _runningDocumentTable = _serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable>();
+            _runningDocumentTableEventTracker = new RunningDocumentTableEventTracker(
+                threadingContext,
+                editorAdaptersFactoryService,
+                _runningDocumentTable,
+                this);
+        }
+
+        public void NavigateToSourceGeneratedFile(Project project, ISourceGenerator generator, string generatedSourceHintName, TextSpan sourceSpan)
+        {
+            // We will create an file name to represent this generated file; the Visual Studio shell APIs imply you can use a URI,
+            // but most URIs are blocked other than file:// and http://; they also get extra handling to attempt to download the file so
+            // those aren't really usable anyways.
+
+            // The file name we generate here is chosen to match the compiler's choice, so the debugger can recognize the files should match.
+            // We use the ProjectId as a parent directory so we can extract that out later too.
+            var generatorType = generator.GetType();
+            var temporaryFilePath = Path.Combine(_temporaryDirectory, $"{project.Id.Id}\\{generatorType.Module.ModuleVersionId}_{generatorType.FullName}_{generatedSourceHintName}");
+            Directory.CreateDirectory(Path.GetDirectoryName(temporaryFilePath));
+            File.WriteAllText(temporaryFilePath, "");
+
+            var openDocumentService = _serviceProvider.GetService<SVsUIShellOpenDocument, IVsUIShellOpenDocument>();
+            var hr = openDocumentService.OpenDocumentViaProject(
```

Comment:
How can we make this work in nexus?  Right now metadata files work I believe because the point to an actual file which liveshare copies over.  Will we be returning the temporary file path to the client?

Reply:
I haven't tried this in Codespaces (yet, which is bad of me) but I suspect this will work as well/poorly as that: in both cases we're creating a temporary file that will get mirrored down.